Add MFA for admin actions docs (#37631)
gravitational/teleport: docs/pages/access-controls/guides/mfa-for-admin-actions.mdx

---
title: MFA for Administrative Actions
description: Require MFA checks to perform administrative actions.
---

Teleport can be configured to require additional multi-factor authentication
checks to perform administrative actions through `tctl`, `tsh`, the Web UI,
Teleport Connect, and any other Teleport client.

Examples of administrative actions include, but are not limited to:

- Resetting or recovering user accounts
- Inviting new users
- Updating cluster configuration resources
- Modifying access management resources
- Approving access requests
- Generating new join tokens
- Impersonation
- Creating new bots for Machine ID

This is an advanced security feature that protects users against compromises of
their on-disk Teleport certificates.

<Notice type="warning">
  When MFA for administrative actions is enabled, user certificates produced
  with `tctl auth sign` will no longer be suitable for automation due to the
  additional MFA checks.

  We recommend using [Machine ID](../../machine-id/getting-started.mdx) to
  issue certificates for automated workflows, which uses role impersonation
  that is not subject to MFA checks.

>   Certificates produced with `tctl auth sign` directly on an Auth service
>   instance using the super-administrative are not subject to MFA checks to support
>   legacy self-hosted setups.

Reviewer: ```suggestion
  Certificates produced with `tctl auth sign` directly on an Auth Service
  instance using the super-administrative are not subject to MFA checks to support
  legacy self-hosted setups.
```

Reviewer: Is "the super-administrative" intended as a modifier for a missing noun (e.g., "role")?

</Notice>

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- (!docs/pages/includes/tctl.mdx!)
- [WebAuthn configured](webauthn.mdx) on this cluster
- Second factor hardware device, such as YubiKey or SoloKey
- A Web browser with [WebAuthn support](
  https://developers.yubico.com/WebAuthn/WebAuthn_Browser_Support/) (if using
  SSH or desktop sessions from the Teleport Web UI).

## Require MFA for administrative actions

MFA for administrative actions is automatically enforced for clusters where
WebAuthn is the only form of second factor allowed.

<Notice type="note">
  In a future major version, Teleport may enforce MFA for administrative actions
  for a wider range of cluster configurations.
</Notice>

<Tabs>

>   <TabItem label="Dynamic resources" scope={["team", "cloud"]}>

Reviewer: ```suggestion
  <TabItem label="Dynamic Resources" scope={["team", "cloud"]}>
```

  Edit the `cluster_auth_preference` resource:

  ```code
  $ tctl edit cap
  ```

  Update the `cluster_auth_preference` definition to include the following content:

  ```yaml
  kind: cluster_auth_preference
  version: v2
  metadata:
    name: cluster-auth-preference
  spec:
    type: local
    # To make webauthn the only form of second factor allowed, set this field to 'webauthn'.
    second_factor: "webauthn"
    webauthn:
      rp_id: example.com
  ```

  Save and exit the file. `tctl` will update the remote definition:

  ```text
  cluster auth preference has been updated
  ```

  </TabItem>
  <TabItem label="Static Config" scope={["oss", "enterprise"]}>

>   Edit the Auth Server's `teleport.yaml` file and restart all Auth Services:

Reviewer: ```suggestion
  Edit the Auth Service's `teleport.yaml` file and restart all Auth Service instances:
```

  `tctl bots add` still requires an MFA check, but the certificates output
  by `tbot` skip the MFA checks.
</Notice>

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- (!docs/pages/includes/tctl.mdx!)
- [WebAuthn configured](webauthn.mdx) on this cluster
- Second factor hardware device, such as YubiKey or SoloKey
- A Web browser with [WebAuthn support](
  https://developers.yubico.com/WebAuthn/WebAuthn_Browser_Support/) (if using
  SSH or desktop sessions from the Teleport Web UI).

## Require MFA for admin actions

MFA for admin actions is automatically enforced for clusters where WebAuthn is
the only allowed form of second factor.

<Notice type="note">
  Teleport may widen the sope of these MFA checks in a future major version to
  cover a wider range of cluster configurations.
</Notice>

<Tabs>
  <TabItem label="Dynamic resources" scope={["team", "cloud"]}>

  Edit the `cluster_auth_preference` resource:

  ```code
  $ tctl edit cap
  ```

  Update the `cluster_auth_preference` definition to include the following content:

  ```yaml
  kind: cluster_auth_preference
  version: v2
  metadata:
    name: cluster-auth-preference
  spec:
    type: local

>     # To enable MFA for admin actinos, set this field to 'webauthn'

Reviewer 2: ```suggestion
    # To enable MFA for admin actions, set this field to 'webauthn'
```

This is a little bit confusing, because there isn't actually a way to "enable MFA for admin actions" - you can only enable other things that indirectly enable the MFA checks.

Author: Fair point, updated to `To make webauthn the only form of second factor allowed, set this field to 'webauthn'.`
